src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs

```diff
@@ -4400,350 +4405,662 @@ internal static IEnumerable<CompletionResult> CompleteFilename(CompletionContext
+                    && wordToComplete[pathStartOffset + 1] is '/' or '\\';
 
-                if (useLiteralPath && LocationGlobber.StringContainsGlobCharacters(wordToComplete))
+                int lastSeparatorIndex = wordToComplete.LastIndexOfAny(Utils.Separators.DirectoryOrDrive);
+                if (lastSeparatorIndex == -1)
                 {
-                    wordToComplete = WildcardPattern.Escape(wordToComplete, Utils.Separators.StarOrQuestion);
+                    // Input is a simple word with no path separators like: "Program Files"
+                    filter = $"{wordToComplete}*";
+                    basePath = ".";
+                    defaultRelativePath = true;
                 }
-
-                if (!defaultRelative && wordToComplete.Length >= 2 && wordToComplete[1] == ':' && char.IsLetter(wordToComplete[0]) && executionContext != null)
+                else
                 {
-                    // We don't actually need the drive, but the drive must be "mounted" in PowerShell before completion
-                    // can succeed.  This call will mount the drive if it wasn't already.
-                    executionContext.SessionState.Drive.GetAtScope(wordToComplete.Substring(0, 1), "global");
+                    if (lastSeparatorIndex + 1 == wordToComplete.Length)
+                    {
+                        // Input ends with a separator like: "./", "filesystem::" or "C:"
+                        filter = "*";
+                        basePath = wordToComplete;
+                    }
+                    else
+                    {
+                        // Input contains a separator, but doesn't end with one like: "C:\Program Fil" or "Registry::HKEY_LOC"
+                        filter = $"{wordToComplete.Substring(lastSeparatorIndex + 1)}*";
+                        basePath = wordToComplete.Substring(0, lastSeparatorIndex + 1);
+                    }
+
+                    if (!inputUsedHomeChar && basePath[0] is not '/' and not '\\')
+                    {
+                        defaultRelativePath = !context.ExecutionContext.LocationGlobber.IsAbsolutePath(wordToComplete, out _);
+                    }
                 }
+            }
 
-                var powerShellExecutionHelper = context.Helper;
-                powerShellExecutionHelper
-                    .AddCommandWithPreferenceSetting("Microsoft.PowerShell.Management\\Resolve-Path")
-                    .AddParameter("Path", wordToComplete + "*");
+            StringConstantType stringType;
+            if (quote.Equals(string.Empty, StringComparison.Ordinal))
+            {
+                stringType = StringConstantType.BareWord;
+            }
+            else if (quote.Equals("\"", StringComparison.Ordinal))
+            {
+                stringType = StringConstantType.DoubleQuoted;
+            }
+            else
+            {
+                stringType = StringConstantType.SingleQuoted;
+            }
```

Comment:
Maybe switch expression is more readable:
```c#
StringConstantType stringType = quote switch
{
    ...
}
```

Reply:
Sure.